usr.sbin/cpuctl/cpuctl.8
.\"	$NetBSD$
.\"
.\" Copyright (c) 2007, 2008, 2012, 2015 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Andrew Doran.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd November 15, 2015
.Dt CPUCTL 8
.Os
.Sh NAME
.Nm cpuctl
.Nd program to control CPUs
.Sh SYNOPSIS
.Nm cpuctl
.Op Fl v
.Ar command
.Op Ar arguments
.Sh DESCRIPTION
The
.Nm
command can be used to control and inspect the state of CPUs in the system.
.Pp
The first argument,
.Ar command ,
specifies the action to take.
Valid commands are:
.Bl -tag -width offline
.It identify Ar cpu Op Ar cpu ...
Output information on the specified CPU's features and capabilities.
Not available on all architectures.
.It list
For each CPU in the system, display the current state and time of the last
state change.
.It offline Ar cpu Op Ar cpu ...
Set the specified CPUs off line.
.Pp
Unbound LWPs (lightweight processes) will not be executed on the CPUs
while it is off line.
Bound LWPs will continue to be executed on the CPUs, and device interrupts
routed to the CPUs will continue to be handled.
A future release of the system may allow device interrupts to be re-routed
away from individual CPUs.
.Pp
At least one CPU in the system must remain on line.
.It online Ar cpu Op Ar cpu ...
Set the specified CPUs on line, making it available to run unbound LWPs.
.It ucode Xo
.Op Ar cpu
.Op Ar file
.Xc
This applies the microcode patch to CPUs.
If
.Ar cpu
is not specified or \-1, all CPUs are updated.
If
.Ar cpu
is \-2, the current CPUs are updated.
The default filename is used if no filename is specified.
The
.Cm identify
command prints the installed version on the specified CPUs.
On success the
.Cm identify
command show different ucode versions before and after this command.
.It intr Ar cpu Op Ar cpu ...
Enable interrupts for the specified CPUs if supported.
.It nointr Ar cpu Op Ar cpu ...
Disable interrupts for the specified CPUs if supported.
.El
.Pp
Valid flags are:
.Bl -tag -width indent
.It Fl v
Be more verbose.
.El
.Sh FILES
.Bl -tag -width /dev/cpuctl -compact
.It Pa /dev/cpuctl
control device
.It Pa /libdata/firmware/x86/amd/
The directory to install the microcode file for AMD CPUs into.
The default filename is
.Pa microcode_amd.bin
for CPU families 0x10 to 0x14.
The default filename is
.Pa microcode_amd_famXXh.bin
where
.Dv XX
is the CPU family starting with 15 (hex).
Get it from
.Lk http://www.amd64.org/microcode.html
.El
.Sh EXAMPLES
Run
.Dl cpuctl identify 0
and you should see something like this:
.Bd -literal -offset indent
cpu0: UCode version: 0x1000080
.Ed
.Pp
After applying the microcode patch with
.Dl cpuctl ucode
you can see with
.Dl cpuctl identify 0
that the patch got applied:
.Bd -literal -offset indent
cpu0: UCode version: 0x1000083
.Ed
.Sh SEE ALSO
.Xr psrset 8 ,
.Xr schedctl 8
.Sh HISTORY
The
.Nm
command first appeared in
.Nx 5.0 .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 7.0 to NetBSD 8.0:
	libc: Update to tzcode2014f. [christos 20140815]
	gmake:	Update to gmake-3.81 in external/gpl2/gmake and remove 
		gnu/dist/gmake (3.80) [christos 20140818]
	module(7): Create MODULAR_DEFAULT_AUTOLOAD config flag to control
		whether module autoloading is on by default.
		[jnemeth 20140824]
	kernel: Option SYMTAB_SPACE has been replaced with COPY_SYMTAB.
		The related changes to src/sys/conf require a rebuild of
		config(1). [joerg 20140824]
	zoneinfo: Import tzdata2014g. [apb 20140831]
	wm(4): Add internal SERDES support for 82575 and newer controllers.
		[msaitoh 20140903]
	or1k: Add new port.  [matt 20140903]
	awindma(4): Add support for Allwinner A10/A20 DMA controller.
		[jmcneill 20140906]
	awinac(4): Add support for Allwinner A10/A20 audio codec.
		[jmcneill 20140906]
	awinrtc(4): Add support for Allwinner A10/A20 real-time clock.
		[jmcneill 20140907]
	awinmmc(4): Use DMA for MMC transfers. [jmcneill 20140908]
	axp20x(4): Add AXP20x PMU driver. [jmcneill 20140909]
	rtsol(8): Removed in favour of dhcpcd. [roy 20140911]
	rtsold(8): Removed in favour of dhcpcd. [roy 20140911]
	dhcpcd(8): Import dhcpcd-6.4.5. [roy 20140918]
	arm: Add support for i.MX6 SoC. [ryo 20140925]
	gpt(8): Completed overhaul, including adding follwing subcommands:
		resize, set, unset, backup, restore, and resizedisk.
		[jnemeth 20140926]
	dhcpcd(8): Import dhcpcd-6.4.7. [roy 20140927]
	mpt(4): added bio(4) support. [jmcneill 20140927]
	gpt(8): added type subcommand to modify partition type.
		[jnemeth 20140927]
	zoneinfo: Import tzdata2014h. [apb 20141004]
	dhcpcd(8): Import dhcpcd-6.5.0. [roy 20141006]
	libc: Update to tzcode2014h. [christos 20141007]
	file(1): Upgraded to 5.20. [christos 20141010]
	arm: Add support for Allwinner A31 SoC. [jmcneill 20141010]
	awiniic(4): Add support for Allwinner A31 SoC. [jmcneill 20141012]
	awinp2wi(4): Add support for Allwinner A31 Push-Pull Two Wire
		Interface. [jmcneill 20141012]
	axp22x(4): Add AXP22x PMU driver. [jmcneill 20141012]
	awinrtc(4): Add support for Allwinner A31 real-time clock.
		[jmcneill 20141012]
	awindma(4): Add support for Allwinner A31 DMA controller.
		[jmcneill 20141013]
	awinac(4): Add support for Allwinner A31 audio codec.
		[jmcneill 20141013]
	motg(4): Add support for Allwinner A10/A20 SoC. [jmcneill 20141015]
	wpa: Import wpa_supplicant and hostapd 2.3. [christos 20141016]
	dhcpcd(8): Import dhcpcd-6.5.1. [roy 20141018]
	OpenSSH: Imported 6.7. [christos 20141018]
	resolvconf(8): Import openresolv-3.6.0 [roy 20141020]
	gcc: Support -fsanitize=address [christos 20141022]
	zoneinfo: Import tzdata2014i. [apb 20141023]
	libc: Update to tzcode2014i. [christos 20141023]
	openpam(3): update to 20140912 (ourouparia) [christos 20141024]
	pppd(8): updated to version 2.4.7. [christos 20141025]
	acpi(4): Updated ACPICA to 20140926. [christos 20141025]
	dhcpcd(8): Import dhcpcd-6.6.0 [roy 20141029]
	resolvconf(8): Import openresolv-3.6.1 [roy 20141029]
	rtsx(4): Add support for Realtek RTS5227/RTL8402/RTL8411/RTL8411B.
		[nonaka 20141029]
	flex(1): Import flex-2.5.37 [christos 20141029]
	iwn(4): Add support for Intel Centrino Wireless-N
		100/105/130/135/2200/2230. From OpenBSD. [nonaka 20141030]
	awge(4): Add support for gigabit ethernet found on various
		Allwinner SOCs. [martin 20141101]
	dhcpcd(8): Import dhcpcd-6.6.1. [roy 20141107]
	awinhdmi(4): Add support for Allwinner A20/A31 HDMI controller.
		[jmcneill 20141110]
	awintcon(4): Add support for Allwinner A20/A31 TV/LCD controller.
		[jmcneill 20141110]
	awindebe(4): Add support for Allwinner A20/A31 Display engine
		backend (DE-BE). [jmcneill 20141110]
	awinhdmiaudio(4): Add support for Allwinner A20/A31 HDMI audio
		controller. [jmcneill 20141111]
	dhcpcd(8): Import dhcpcd-6.6.2. [roy 20141114]
	awinir(4): Add support for Allwinner A20/A31 IR receiver.
		[jmcneill 20141115]
	zoneinfo: Import tzdata2014j. [apb 20141116]
	libc: Change arc4random(3) to use ChaCha20 and per-thread state.
		[riastradh 20141116]
	libc: Update to tzcode2014j. [christos 20141117]
	libpcap: Import 1.6.2. [christos 20141119]
	tcpdump(8): Import 4.6.2. [christos 20141119]
	dhcpcd(8): Import dhcpcd-6.6.4. [roy 20141126]
	dhcpcd(8): Import dhcpcd-6.6.5. [roy 20141209]
	bind: Import version 9.10.1-P1. [christos 20141209]
	dhcpcd(8): Import dhcpcd-6.6.6. [roy 20141217]
	powerpc: Switch to GCC 4.8. [matt 20141218]
	dhcpcd(8): Import dhcpcd-6.6.7. [roy 20141219]
	ntp: Import ntp 4.2.8. [christos 20141219]
	midirecord(1): Add program to record MIDI files.  [mrg 20141229]
	file: update to 5.22 security limits [christos 20150102]
	byacc: update to 20141128 for more bison compat [christos 20150103]
	mgx: driver for Southland Media Systems (now Quantum 3D) MGX SBus
		graphics devices [macallan 20150106]
	gcc: Import GCC 4.8.4.  [mrg 20150109]
	openssl: Import openssl 1.0.1k [spz 20150113]
	am-utils: Upgrade to 6.2; adds nfsv4, lustre (which we can't use)
		nfs3 rpc support for amfs [christos 20150117]

usr.sbin/cpuctl/cpuctl.c
/*	$NetBSD$	*/

/*-
 * Copyright (c) 2007, 2008, 2009, 2012, 2015 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Andrew Doran.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef lint
#include <sys/cdefs.h>
__RCSID("$NetBSD$");
#endif /* not lint */

#include <sys/param.h>
#include <sys/ioctl.h>
#include <sys/uio.h>
#include <sys/cpuio.h>

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <paths.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <unistd.h>
#include <util.h>
#include <time.h>
#include <sched.h>
#include <stdbool.h>

#include "cpuctl.h"

static u_int	getcpuid(char *);
__dead static void	usage(void);

static void	cpu_identify(char **);
static void	cpu_list(char **);
static void	cpu_offline(char **);
static void	cpu_online(char **);
static void	cpu_intr(char **);
static void	cpu_nointr(char **);
static void	cpu_ucode(char **);

static struct cmdtab {
	const char	*label;
	bool	takesargs;
	bool	argsoptional;
	void	(*func)(char **);
} const cpu_cmdtab[] = {
	{ "identify",	true,  false, cpu_identify },
	{ "list",	false, false, cpu_list },
	{ "offline",	true,  false, cpu_offline },
	{ "online",	true,  false, cpu_online },
	{ "intr",	true,  false, cpu_intr },
	{ "nointr",	true,  false, cpu_nointr },
	{ "ucode",	true,  true,  cpu_ucode },
	{ NULL,		false, false, NULL },
};

static int	fd;
int		verbose;

int
main(int argc, char **argv)
{
	const struct cmdtab *ct;
	int ch;

	while ((ch = getopt(argc, argv, "v")) != -1)
		switch (ch) {
		case 'v':
			verbose = 1;
			break;
		default:
			usage();
		}
	argc -= optind;
	argv += optind;
	if (argc < 1)
		usage();

	if ((fd = open(_PATH_CPUCTL, O_RDWR)) < 0)
		err(EXIT_FAILURE, _PATH_CPUCTL);

	for (ct = cpu_cmdtab; ct->label != NULL; ct++) {
		if (strcmp(argv[0], ct->label) == 0) {
			if (!ct->argsoptional &&
			    ((ct->takesargs == 0) ^ (argv[1] == NULL)))
			{
				usage();
			}
			(*ct->func)(argv + 1);
			break;
		}
	}

	if (ct->label == NULL)
		errx(EXIT_FAILURE, "unknown command ``%s''", argv[0]);

	close(fd);
	exit(EXIT_SUCCESS);
	/* NOTREACHED */
}

static void
usage(void)
{
	const char *progname = getprogname();

	fprintf(stderr, "usage: %s identify cpuno\n", progname);
	fprintf(stderr, "       %s list\n", progname);
	fprintf(stderr, "       %s offline cpuno\n", progname);
	fprintf(stderr, "       %s online cpuno\n", progname);
	fprintf(stderr, "       %s intr cpuno\n", progname);
	fprintf(stderr, "       %s nointr cpuno\n", progname);
	fprintf(stderr, "       %s ucode [cpuno] [file]\n", progname);
	exit(EXIT_FAILURE);
	/* NOTREACHED */
}

static void
cpu_online(char **argv)
{
	cpustate_t cs;

	for (; *argv; argv++) {
		cs.cs_id = getcpuid(*argv);
		if (ioctl(fd, IOC_CPU_GETSTATE, &cs) < 0)
			err(EXIT_FAILURE, "IOC_CPU_GETSTATE");
		cs.cs_online = true;
		if (ioctl(fd, IOC_CPU_SETSTATE, &cs) < 0)
			err(EXIT_FAILURE, "IOC_CPU_SETSTATE");
	}
}

static void
cpu_offline(char **argv)
{
	cpustate_t cs;

	for (; *argv; argv++) {
		cs.cs_id = getcpuid(*argv);
		if (ioctl(fd, IOC_CPU_GETSTATE, &cs) < 0)
			err(EXIT_FAILURE, "IOC_CPU_GETSTATE");
		cs.cs_online = false;
		if (ioctl(fd, IOC_CPU_SETSTATE, &cs) < 0)
			err(EXIT_FAILURE, "IOC_CPU_SETSTATE");
	}
}

static void
cpu_intr(char **argv)
{
	cpustate_t cs;

	for (; *argv; argv++) {
		cs.cs_id = getcpuid(*argv);
		if (ioctl(fd, IOC_CPU_GETSTATE, &cs) < 0)
			err(EXIT_FAILURE, "IOC_CPU_GETSTATE");
		cs.cs_intr = true;
		if (ioctl(fd, IOC_CPU_SETSTATE, &cs) < 0)
			err(EXIT_FAILURE, "IOC_CPU_SETSTATE");
	}
}

static void
cpu_nointr(char **argv)
{
	cpustate_t cs;

	for (; *argv; argv++) {
		cs.cs_id = getcpuid(*argv);
		if (ioctl(fd, IOC_CPU_GETSTATE, &cs) < 0)
			err(EXIT_FAILURE, "IOC_CPU_GETSTATE");
		cs.cs_intr = false;
		if (ioctl(fd, IOC_CPU_SETSTATE, &cs) < 0) {
			if (errno == EOPNOTSUPP) {
				warnx("interrupt control not supported on "
				    "this platform");
			} else
				err(EXIT_FAILURE, "IOC_CPU_SETSTATE");
		}
	}
}

static void
cpu_ucode(char **argv)
{
	int error;
	struct cpu_ucode uc;
	unsigned long id = 0; /* gcc */
	char *ep;
	cpuset_t *cpuset;

	uc.cpu_nr = -1;
	if (argv[0] != NULL) {
		id = strtoul(argv[0], &ep, 0);
		if (id != ULONG_MAX && *ep == '\0') {
			uc.cpu_nr = id;
			argv++;
		}
	}
	if (argv[0] != NULL)
		strlcpy(uc.fwname, argv[0], sizeof(uc.fwname));
	else
		memset(uc.fwname, '\0', sizeof(uc.fwname));

	error = ucodeupdate_check(fd, &uc);
	if (error)
		errx(EXIT_FAILURE, "unsupported");

	if (uc.cpu_nr == CPU_UCODE_CURRENT_CPU) {
		cpuset = cpuset_create();
		if (cpuset == NULL)
			err(EXIT_FAILURE, "cpuset_create");
		cpuset_zero(cpuset);
		cpuset_set(id, cpuset);
		if (_sched_setaffinity(0, 0, cpuset_size(cpuset), cpuset) < 0) {
			err(EXIT_FAILURE, "_sched_setaffinity");
		}
		cpuset_destroy(cpuset);
	}
	error = ioctl(fd, IOC_CPU_UCODE_APPLY, &uc);
	if (error < 0) {
		if (uc.fwname[0])
			err(EXIT_FAILURE, "%s", uc.fwname);
		else
			err(EXIT_FAILURE, "IOC_CPU_UCODE_APPLY");
	}
}


static void
cpu_identify(char **argv)
{
	char name[32];
	unsigned int id, np;
	cpuset_t *cpuset;

	np = sysconf(_SC_NPROCESSORS_CONF);
	for (; *argv; argv++) {
		id = getcpuid(*argv);
		snprintf(name, sizeof(name), "cpu%u", id);

		if (np != 1) {
			cpuset = cpuset_create();
			if (cpuset == NULL)
				err(EXIT_FAILURE, "cpuset_create");
			cpuset_zero(cpuset);
			cpuset_set(id, cpuset);
			if (_sched_setaffinity(0, 0, cpuset_size(cpuset), cpuset) < 0) {
				if (errno == EPERM) {
					printf("Cannot bind to target CPU.  Output "
					    "may not accurately describe the target.\n"
					    "Run as root to allow binding.\n\n");
				} else { 
					err(EXIT_FAILURE, "_sched_setaffinity");
				}
			}
			cpuset_destroy(cpuset);
		}
		identifycpu(fd, name);
	}
}

static u_int
getcpuid(char *arg)
{
	char *argp;
	u_int id;
	long np;

	id = (u_int)strtoul(arg, &argp, 0);
	if (*argp != '\0')
		usage();

	np = sysconf(_SC_NPROCESSORS_CONF);
	if (id >= (u_long)np)
		errx(EXIT_FAILURE, "Invalid CPU number");

	return id;
}

static void
cpu_list(char **argv)
{
	const char *state, *intr;
	cpustate_t cs;
	u_int cnt, i;
	time_t lastmod;
	char ibuf[16], *ts;
	
	if (ioctl(fd, IOC_CPU_GETCOUNT, &cnt) < 0)
		err(EXIT_FAILURE, "IOC_CPU_GETCOUNT");

	printf(
"Num  HwId Unbound LWPs Interrupts Last change              #Intr\n"
"---- ---- ------------ ---------- ------------------------ -----\n");

	for (i = 0; i < cnt; i++) {
		cs.cs_id = i;
		if (ioctl(fd, IOC_CPU_GETSTATE, &cs) < 0)
			err(EXIT_FAILURE, "IOC_CPU_GETSTATE");
		if (ioctl(fd, IOC_CPU_MAPID, &cs.cs_id) < 0)
			err(EXIT_FAILURE, "IOC_CPU_MAPID");
		if (cs.cs_online)
			state = "online";
		else
			state = "offline";
		if (cs.cs_intr)
			intr = "intr";
		else
			intr = "nointr";
		if (cs.cs_intrcnt == 0)
			strcpy(ibuf, "?");
		else
			snprintf(ibuf, sizeof(ibuf), "%d", cs.cs_intrcnt - 1);

		lastmod = (time_t)cs.cs_lastmod |
		    ((time_t)cs.cs_lastmodhi << 32);
		ts = asctime(localtime(&lastmod));
		ts[strlen(ts) - 1] = '\0';
		printf("%-4d %-4x %-12s %-10s %s %-5s\n",
		   i, cs.cs_hwid, state,
		   intr, ts, ibuf);
	}
}

int
aprint_normal(const char *fmt, ...)
{
	va_list ap;
	int rv;

	va_start(ap, fmt);
	rv = vfprintf(stdout, fmt, ap);
	va_end(ap);

	return rv;
}
__strong_alias(aprint_verbose,aprint_normal)
__strong_alias(aprint_error,aprint_normal)

int
aprint_normal_dev(const char *dev, const char *fmt, ...)
{
	va_list ap;
	int rv;

	printf("%s: ", dev);
	va_start(ap, fmt);
	rv = vfprintf(stdout, fmt, ap);
	va_end(ap);

	return rv;
}
__strong_alias(aprint_verbose_dev,aprint_normal_dev)
__strong_alias(aprint_error_dev,aprint_normal_dev)
